### ISSUE TITLE
How to transform output prediction back to garment

### ISSUE THREAD

**Issue by shawkontzu** [HugoBA92/DeePSD#3]
Hi, please indicate how to transform the output of model's prediction [pred, body] back to deformed garment object, there seems to be no testing script from the repository.

Thank you

**Comment by hbertiche** [comment]
Hi,

The output of the model [pred, body] are two tensorflow matrices with the vertex locations for the garment/outfit and the body,
pred: N x 3
body: 6890 x 3
you can call the model with 'with_body=False' if you do not want it to compute SMPL, in this case, 'body' will be None

then, in 'IO.py' there are functions for reading and writing 3D data,
if you want to store a single prediction, the best option is 'writeOBJ(file, verts, faces)',
where 'file' is the name of the file that shall be created,  'verts' must be a N x 3 numpy matrix and F a list of faces,

if you want to store animation data, you should use 'writePC2(file, verts)',
again, 'file' is the name that will be used for saving, and in this case 'verts' is a T x N x 3 numpy matrix, where 'T' corresponds to the number of frames in the animation,

OBJ and PC2 files are compatible with Blender, so you should be able to visualize your predictions there,
I believe these data formats are also compatible with other 3D tools, it is not tested, though,
also, when importing an OBJ in Blender, make sure to use the option 'Keep Vertex Order', otherwise, when you apply the animation data (PC2) to this object, the topology will be messy (wrong vertex order)

**Comment by shawkontzu** [comment]
@HugoBA92 
Hi, thank you for the reply! As you mentioned the output prediction only has vertices as output, but saving to obj file requires 3D faces F as input. How to create faces from prediction? 
Thank you

**Comment by hbertiche** [comment]
Hi,

no need to create faces for the predictions, you can use the faces of the input template garment,
the object 'batch' in 'train.py' (ln. 73) has a field called 'faces' as a Nx3 array,

nonetheless, note that, given that each sample has different topology and dimensionality, during training, it is convinient to fuse all the outfits in the batch into a single mesh,
in order to separate them into independent meshes again, you can use the information at batch['indices'],
nonetheless, for the faces, this is not straightforward,
for a simpler solution, I recommend using a batch_size=1, then, the data within 'batch' will correspond to a single outfit,

**Comment by shawkontzu** [comment]
@HugoBA92 Thank you for the reply.

I have tried to generate a few garments using the same smpl params but the output garment objects seem to be quite different in terms of corresponding body shapes. 
<img width="395" alt="Screen Shot 2021-08-05 at 10 32 24 AM" src="https://user-images.githubusercontent.com/81662273/128397840-ad959302-553b-4675-91da-88b2d020c201.png">
<img width="418" alt="Screen Shot 2021-08-05 at 10 32 13 AM" src="https://user-images.githubusercontent.com/81662273/128397876-58c6c309-1bd9-48c8-a2ab-f5a2ef5b6cef.png">

Take these 2 results for example, these are 2 training data that were used for training the deepsd model, I have changed the shape, pose input params in testing and kept the rest the same (including the preprocessing data 'template' and 'laplacian') for DeePSD prediction. 

Could you please let me know how to generalize DeePSD prediction to different body shapes if I use the same CLOTH3D garment templates?

**Comment by hbertiche** [comment]
hi,

DeePSD does not resize outfits, it is focused on animation only,
DeePSD predicts blend weights and pose space deformations for a given template outfit, it does not change this template,
the model takes 'shape' as input only to compute the forward kinematics w.r.t. SMPL skeleton (SMPL needs shape to compute joint locations),

if you are interested in outfit resizing, in the CLOTH3D paper we proposed a simple automatic algorithm for this, no deep learning, so it can be applied to any outfit without any restriction,
https://arxiv.org/pdf/1912.02792.pdf
in sec.3.2, in the paragraph 'Resizing',

alternatively, in a recent work, we propose an unsupervised approach for garment animation, we show it can also be used for automatic resizing with control over tightness, note that it is outfit-specific, it does not generalize to new outfits,
https://arxiv.org/pdf/2012.11310.pdf
sec.5.6 describes how to adapt the main methodology for resizing,
the code for this will be public soon,

once you have a resized outfit to the desired body shape, you just need to feed DeePSD with the template and corresponding body shape

**Comment by shawkontzu** [comment]
Hi @HugoBA92, thank you for this information. I am still confused about how CLOTH3D does garment resizing after reading [CLOTH3D paper](https://arxiv.org/pdf/1912.02792.pdf) and there seems to be no CLOTH3D repository showing how the dataset was built. Could you please share the script you used for garment resizing? Or provide a more detailed explanation on how to resize the garment based on existing data we have from CLOTH3D dataset? 

I am confused about the part using betas[0], betas[1] for rescale the garment template. For smpl default template mesh betas equals to np.zeros((1, 10)), reshaping the smpl model uses (betas * shapedirs).sum(dim=-1) + v_template where v_template is the vertices for template smpl mesh, this is adding a displacement to every vertex on smpl template. How are the betas[0] and betas[1] information applied to the rescale of garment? 

Thank you very much!

**Comment by impanel** [comment]
Hi guys, 

thanks for this thread and thanks @HugoBA92 for this great work. It's been incredibly useful and I was able to train the model and get some good prediction results. But just as @shawkontzu I've been struggling with resizing the garment to fit it onto my SMPL mesh.
I've read the paper's resizing paragraph and understand that I need to rescale the garment based on the first 2 SMPL beta parameters (size, fatness). But I'm struggling to understand the displacement part. Am I supposed to find the closest vertex between garment and body and displace the garment vertex using the distance between the two? Doesn't that just create a garment that fits tight to the body.

It would be great if you could provide a more detailed pointer on how to best resize the predicted garment. 

Thanks

**Comment by hbertiche** [comment]
hi,

sorry for the delayed response guys, 
we havent released CLOTH3D code, we would like to do this in the future, but unfortunately, it is not a priority right now, so I do not know when this will happen,

I will try now to explain in more detail how to resize garments as explained for CLOTH3D,
note that this would allow resizing garments in rest pose, later DeePSD can be used to animate those,
the idea to resize these garments is very similar to SMPL shaping,
for SMPL, we have a template of 6890 vertices (6890 x 3 matrix),
then, for reshaping, it has a matrix like 6890 x 3 x 10, and shape is expressed as a 10 float array, thus, combining the matrix and the shape params (6890 x 3 x 10) x (10) => 6890 x 3, you get displacement for each vertex,
we want to do the same for a given outfit in rest pose,
the outfit has N vertices, that is, a matrix like N x 3,
then, to use the same reshaping mechanism, we need a matrix N x 3 x 10,
to obtain this matrix, we compute the closest body vertex for each cloth vertex (body and outfit must be aligned),
then, you can have an N x 3 x 10 matrix by transfering the per-vertex 3x10 matrices from the 6890x3x10 SMPL matrix (shapedirs),
(e.g.: if the 'j'-th SMPL vertex is the closest to the 'i'-th cloth vertex, then, "outfit_shapedirs[i] = smpl_shapedirs[j]")
finally, to avoid carrying body geometry details to the outfit, before using 'outfit_shapedirs', smooth it using the outfit laplacian, this means that each vertex 'shapedir' should be the average of their neighbours, do this around 20-100 times,
once this is done, you can reshape the outfit in the same way that SMPL is reshaped, just remember to take into the account the shape parameters of the original outfit template (it is easier if template is for 0-shaped bodies),
another consideration, smoothing meshes yields volume loss, in this case, we smooth 'shapedirs', the effect it will have is that re-shaping will have 'lower' effect, then, for very big bodies, skin-to-cloth collisions might appear, I recommend to apply a collision post-processing to the reshaped template outfit

**Comment by shawkontzu** [comment]
Hi @HugoBA92 , thank you so much for the explanation! This is very helpful!

I have a question regarding smoothing the outfit_shapedirs. I understand now that the outfit_shapedirs is generated using the shapedirs of nearest vertex in the smpl mesh, and I understand the idea smoothing the outfit_shapedirs (Nx3x10, N=number of vertices in the garment) before adding garment displacement but not smoothing the garment directly. I am still confused about how to "smooth outfit_shapedirs using the outfit laplacian"? Seems like the laplacian.npz file stores the "laplacian matrix" for each garment, it is a NXN matrix (N=number of vertices in the garment), where for each vertex v and its neighbor vertex v', laplacianMatrix[v][v'] = 1/n (n=number of neighboring vertices for v). Could you please help me to understand how to "smooth" this NX3X10 outfit_shapedirs array by using this NXN laplacian matrix? 

Another question is about "take into the account the shape parameters of the original outfit template", since all garment from CLOTH3D dataset are reshaped garments, how do we consider the original shape parameters we got from garment info.mat file in this resizing process? 

Thank you!

**Comment by hbertiche** [comment]
to use the laplacians for smoothing, for a mesh with vertices V (N x 3),
V_smooth = laplacian x V
with a matrix product, 
for the shapedirs you can do the same thing, for simplicity, I first reshape shapedirs, from N x 3 x 10 to N x 30, then apply the laplacian, and reshape again to N x 3 x 10,

then, to take into account the shape parameters of the template, you have to assume that this template is the result of applying shape deformations to an "imaginary" un-shaped template for this outfit,
T = T_0 + D*shape
where 'T' is the template outift, 'T_0' is the imaginary un-shaped template, 'D' are the obtained shapedirs and 'shape' is the shape for that outfit,
from here, you can compute T_0 and then apply the same formulation for any other shape,

**Comment by shawkontzu** [comment]
Hi @HugoBA92 Thank you so much for this information. I have made a script to resize the garment we get from CLOTH3D back to its template garment. Could you please let me know if this makes sense to you?

resize shaped garment back to its template garment:
1. Load CLOTH3D shaped garment object -> shaped_garment
2. Read betas from info.mat included in the garment directory -> betas
3. Use betas from step 2 to generate a smpl object -> shaped_smpl_object
4. Align shaped_garment from step1 and shaped_smpl_object from step 3, find closest vertex pairs between shaped_garment and shaped_smpl_object 
5. Load shapedirs from smpl model file -> shapedirs
6. Use the closest neighbor vertex pairs from step 4 to generate garment shapedirs -> garment_shapedirs
7. Laplacian smoothing on garment_shapedirs to get smoothed garment_shapedirs -> smoothed_garment_shapedirs
8. Getting the template garment should be the reverse of CLOTH3D resizing, which instead of adding the displacement, it should be shaped_garment to substract the displacement and get the template garment. -> template_garment = shaped_garment - smoothed_garment_shapedirs * betas

Please let me know if this is the correct steps for getting the template garment. Then resizing would be similar to the above steps except that we start from a template garment, and last step would be adding the calculated displacements to the template garment.

**Comment by hbertiche** [comment]
hi,

it seems correct,
please take into account that the rest pose we use in CLOTH3D is not all zeros, but the one you will find at 'DeePSD/values.py' as 'rest_pose'
also note that, while generating CLOTH3D dataset, we applied this process, but always to zero-shaped template garments, 
it is likely that estimating the 'imaginary' neutral shape template introduces some error,

**Comment by shawkontzu** [comment]
Hi @HugoBA92 , thank you for the clarification.
I created a blender python script based on our previous discussion about resizing CLOTH3D data back to its template garment. Here's an example for resizing garment 00008 (jumpsuit) to its template garment. 
<img width="945" alt="Screen Shot 2021-10-07 at 10 31 54 AM" src="https://user-images.githubusercontent.com/81662273/136434994-5f6c8089-f72c-4778-a54f-0f73432d9f54.png">
<img width="905" alt="Screen Shot 2021-10-07 at 10 30 44 AM" src="https://user-images.githubusercontent.com/81662273/136435019-bb88dce0-361c-45c2-bc36-1d49d8064fbe.png">
<img width="585" alt="Screen Shot 2021-10-07 at 10 30 55 AM" src="https://user-images.githubusercontent.com/81662273/136435045-512b858f-32e0-40e5-aa5e-3d2f23b87d93.png">

The blue jumpsuit on the left is the cloth3D shaped garment mesh, and the smpl object on the left is the smpl mesh I created using info.mat for garment 00008. The garment seems to fit perfectly on the shaped smpl mesh. The red jumpsuit on the right is the resized template garment using the approach above, it should fit to the template smpl mesh which I generated using default pose and 0 betas. I would assume the garment template should fit smpl template but the result shows that there's distinctive misalignment between these two. The garment is resized because the red jumpsuit is smaller comparing to the blue jumpsuit, but the shape features including the big belly from blue jumpsuit seems to persist in the resized garment. 

Here's the blender python script that I used for resizing cloth3D garment back to its template garment, please let me know if you see any issues with the resizing process.
[https://github.com/shawkontzu/garment_resizing/blob/main/cloth3D_garment_resizing.py](url)

Thank you

**Comment by hbertiche** [comment]
hi,

I would recommend starting simple,
use a sample with neutral shape, and then try to shape it along its outfit to another shape,
this should be easier than un-shaping, once you got this working, you can try the opposite,

it is unlikely to find a sample with 0 betas,
for this, I would suggest to pick the outfit of a sample with betas close to 0 and manually edit the outfit with Blender to align it with the neutral body shape,
alternatively, the cloth modifier might help you with this,

I see you posted your script,
unfortunately, as you may understand, it would require a significant effort from my part to debug this for you,
thus, I will not be able to help you with this,
